upstairs/src/live_repair.rs

```diff
@@ -0,0 +1,6113 @@
+//
+// Construct the correct job and return that to the caller.
+// everything is good, then just submit a NoOp job.
+#[allow(clippy::too_many_arguments)]
+fn repair_or_noop(
+    ds: &mut Downstairs,
+    extent: usize,
+    repair_id: u64,
+    repair_deps: Vec<u64>,
+    gw_repair_id: u64,
+    impacted_blocks: ImpactedBlocks,
+    source: u8,
+    repair: Vec<u8>,
+) -> DownstairsIO {
+    assert!(repair.len() < 3);
+
+    let mut need_repair = Vec::new();
+    debug!(ds.log, "Get repair info for {} source", source);
+    let good_ei = ds.repair_info.remove(&source).unwrap();
+    for broken_extent in repair.iter() {
+        debug!(ds.log, "Get repair info for {} bad", broken_extent);
+        let repair_ei = ds.repair_info.remove(broken_extent).unwrap();
+
+        let repair = if repair_ei.dirty {
+            true
+        } else if repair_ei.generation != good_ei.generation {
+            true
+        } else {
+            repair_ei.flush_number != good_ei.flush_number
+        };
+
+        if repair {
+            need_repair.push(*broken_extent);
+        }
+    }
+
+    // Now that we have consumed the contents, be sure to clear
+    // out anything we did not look at.  There could be something left
+    ds.repair_info = HashMap::new();
+
+    if need_repair.is_empty() {
+        info!(ds.log, "No repair needed for extent {}", extent);
+        for cid in repair.iter() {
+            ds.extents_confirmed[*cid as usize] += 1;
+        }
+        create_noop_io(repair_id, repair_deps, gw_repair_id, impacted_blocks)
+    } else {
+        info!(
+            ds.log,
+            "Repair for extent {} s:{} d:{:?}", extent, source, need_repair
+        );
+        for cid in repair.iter() {
+            ds.extents_repaired[*cid as usize] += 1;
+        }
+        let repair_address = ds.repair_addr(source);
+
+        create_repair_io(
+            repair_id,
+            repair_deps,
+            gw_repair_id,
+            extent,
+            impacted_blocks,
+            repair_address,
+            source,
+            need_repair,
+        )
+    }
+}
+
+// Build the list of dependencies for a live repair job.
+// Because we need all three repair jobs to happen lock step, we
+// have to prevent any IO from hitting the same extent, which means
+// any IO going to our ImpactedBlocks (the whole extent) must finish
+// first (or come after) our job.
+fn deps_for_live_repair(
+    ds: &Downstairs,
+    impacted_blocks: ImpactedBlocks,
+    close_id: u64,
+) -> Vec<u64> {
+    let num_jobs = ds.ds_active.keys().len();
+    let mut deps: Vec<u64> = Vec::with_capacity(num_jobs);
+
+    // Search backwards in the list of active jobs, stop at the
+    // last flush
+    for job_id in ds
+        .ds_active
+        .keys()
+        .sorted()
+        .collect::<Vec<&u64>>()
+        .iter()
+        .rev()
+    {
+        let job = &ds.ds_active[job_id];
+
+        // If we scheduled a job that required reservation of repair IDs,
```

Comment:
i dont really understand what this this comment is saying. so we scheduled a job. we had to reserve repair job IDs. we're figuring out what the repair-job depends on. so we dont want the repair job to depend on the job that depends on the repair job? sorry im a bit confuzzled.

Reply:
You are correct in what you state here, but I'll improve the comment to hopefully make it clearer.

Reply:
Updated comment